ts/src/program/namespace/types.ts

```diff
@@ -0,0 +1,139 @@
+import { PublicKey } from "@solana/web3.js";
+import BN from "bn.js";
+import { Idl } from "src";
+import {
+  IdlField,
+  IdlInstruction,
+  IdlType,
+  IdlTypeDef,
+  IdlTypeDefTyEnum,
+  IdlTypeDefTyStruct,
+} from "../../idl";
+import { Accounts, Context } from "../context";
+
+/**
+ * All instructions for an IDL.
+ */
+export type AllInstructions<IDL extends Idl> = IDL["instructions"][number];
+
+/**
+ * Returns a type of instruction name to the IdlInstruction.
+ */
+export type InstructionMap<I extends IdlInstruction> = {
+  [K in I["name"]]: I & { name: K };
+};
+
+/**
+ * Returns a type of instruction name to the IdlInstruction.
+ */
+export type AllInstructionsMap<IDL extends Idl> = InstructionMap<
+  AllInstructions<IDL>
+>;
+
+/**
+ * All accounts for an IDL.
+ */
+export type AllAccounts<IDL extends Idl> = IDL["accounts"] extends undefined
+  ? IdlTypeDef
+  : NonNullable<IDL["accounts"]>[number];
+
+/**
+ * Returns a type of instruction name to the IdlInstruction.
+ */
+export type AccountMap<I extends IdlTypeDef> = {
+  [K in I["name"]]: I & { name: K };
+};
+
+/**
+ * Returns a type of instruction name to the IdlInstruction.
+ */
+export type AllAccountsMap<IDL extends Idl> = AccountMap<AllAccounts<IDL>>;
+
+export type MakeInstructionsNamespace<
+  IDL extends Idl,
+  I extends IdlInstruction,
+  Ret,
+  Mk extends { [M in keyof InstructionMap<I>]: unknown } = {
+    [M in keyof InstructionMap<I>]: unknown;
+  }
+> = {
+  [M in keyof InstructionMap<I>]: InstructionContextFn<
+    IDL,
+    InstructionMap<I>[M],
+    Ret
+  > &
+    Mk[M];
+};
+
+export type InstructionContextFn<
+  IDL extends Idl,
+  I extends AllInstructions<IDL>,
+  Ret
+> = (...args: InstructionContextFnArgs<IDL, I>) => Ret;
+
+export type InstructionContextFnArgs<
```

Comment:
This doesn't typecheck properly for types with no args and no account context. E.g., `program.rpc.initialize()`--when you run `anchor init <name>` to generate a new workspace.

Reply:
We need to make the Context here optional.

Reply:
Punting this. Instead ts clients will just require the Context object.